### ISSUE TITLE
[ENH] Optimize round() calls on C floats and C doubles

### ISSUE THREAD

**Issue by bbb23exposed** [cython/cython#6447]
### Is your feature request related to a problem? Please describe.

```cython
cdef float test(float value):
    return round(value)
```

and

```cython
cdef double test(double value):
    return round(value)
```

First convert the value to a PyObject to call round() and then converts it back. These should be optimized to C function calls.

**Comment by scoder** [comment]
I don't think there is a public C-API for Python's random module. And we cannot just use the C functions of the same name because they use different state internally.

**Comment by bbb23exposed** [comment]
> I don't think there is a public C-API for Python's random module. And we cannot just use the C functions of the same name because they use different state internally.

?

What does this have to do with random?

**Comment by scoder** [comment]
Ah, sorry. Strike that. For some reason, I misread your request.

**Comment by scoder** [comment]
Python's implementation of `round()` isn't quite as simple as you might think:
https://github.com/python/cpython/blob/c35b33bfb7c491dfbdd40195d70dcfc4618265db/Objects/floatobject.c#L938-L1108

You wouldn't want
```python
round(3.14)
```
and
```python
x: cython.double = 3.14
round(x)
```
to give different results for some floating point values.

**Comment by da-woods** [comment]
> Seems like that only matters when the full value can't be contained in the output, but my patch only operates on C types so that shouldn't be an issue. If the output is a python float or python int, it still uses the python round().

I'm not sure I believe this argument at all (or actually understand what you're saying).

A Python `float` is just a C `double`  stored in a Python object. Therefore `round` should behave identically for a Python `float` and a C `double` because the underlying data is *exactly the same*.

The results are provable not the same. Here's a simple program that just cycles through a few double representations (essentially changing one bit at a time) until it finds one that's different.

```
from libc.math cimport round as cround
from libc.stdint cimport uint64_t

def run():
    cdef uint64_t i

    cdef double start = 0.4999999999
    cdef uint64_t start_repr = (<uint64_t*>&start)[0]

    for i in range(start_repr, 2**64-1):
        i_as_double_ptr = <double*>(&i)
        i_as_double = i_as_double_ptr[0]

        if i % 1000000 == 0:
            print(i, i_as_double)

        crounded = cround(i_as_double)
        rounded = round(i_as_double)
        if crounded != rounded:
            print("Failed:", repr(i_as_double), crounded, rounded)
            break
```

**Comment by da-woods** [comment]
Although frankly that program is overkill to prove that `round(0.5)` is different

**Comment by bbb23exposed** [comment]
> > Seems like that only matters when the full value can't be contained in the output, but my patch only operates on C types so that shouldn't be an issue. If the output is a python float or python int, it still uses the python round().
> 
> I'm not sure I believe this argument at all (or actually understand what you're saying).
> 
> A Python `float` is just a C `double` stored in a Python object. Therefore `round` should behave identically for a Python `float` and a C `double` because the underlying data is _exactly the same_.
> 
> The results are provable not the same. Here's a simple program that just cycles through a few double representations (essentially changing one bit at a time) until it finds one that's different.
> 
> ```
> from libc.math cimport round as cround
> from libc.stdint cimport uint64_t
> 
> def run():
>     cdef uint64_t i
> 
>     cdef double start = 0.4999999999
>     cdef uint64_t start_repr = (<uint64_t*>&start)[0]
> 
>     for i in range(start_repr, 2**64-1):
>         i_as_double_ptr = <double*>(&i)
>         i_as_double = i_as_double_ptr[0]
> 
>         if i % 1000000 == 0:
>             print(i, i_as_double)
> 
>         crounded = cround(i_as_double)
>         rounded = round(i_as_double)
>         if crounded != rounded:
>             print("Failed:", repr(i_as_double), crounded, rounded)
>             break
> ```

Okay, this should be fixed now.

**Comment by da-woods** [comment]
> ndigits isn't set, then C round() is called and the only thing else that happens is the adjustment for even/non-even.

Yes - missed that on first look but you're right - the odd/even is the only difference
